examples/natmod/stm32wb55_transparent_vcp/stm32wb55_transparent.c

```diff
@@ -0,0 +1,177 @@
+#include "py/dynruntime.h"
+#include "py/binary.h"
+#include "py/objarray.h"
+
+#include "stm32wb55_local_commands.h"
+
+// Don't enable this if stdio is used for transp comms.
+#define DEBUG_printf(...)  // mp_printf(&mp_plat_print, "rfcore_transp: " __VA_ARGS__)
+
+#define STATE_IDLE 0
+#define STATE_NEED_LEN 1
+#define STATE_IN_PAYLOAD 2
+
+#define HCI_KIND_BT_CMD (0x01) // <kind=1><?><?><len>
+#define HCI_KIND_BT_ACL (0x02) // <kind=2><?><?><len LSB><len MSB>
+#define HCI_KIND_BT_EVENT (0x04) // <kind=4><op><len><data...>
+#define HCI_KIND_VENDOR_RESPONSE (0x11)
+#define HCI_KIND_VENDOR_EVENT (0x12)
+#define HCI_KIND_LOCAL_CMD (0x20) // Used by STM32CubeMonitor to query the device
+#define HCI_KIND_LOCAL_RSP (0x21)
+
+
+// Callback python function, can be used to provide feedback on each comm, eg. blink LED.
+void mpy_run_callback(mp_obj_t callback, bool on) {
+    if (callback != mp_const_none) {
+        mp_call_function_n_kw(callback, 1, 0, (mp_obj_t[]) {(on) ? mp_const_true : mp_const_false});
+    }
+}
+
+
+mp_obj_t mp_stream_write(mp_obj_t stream_out, const void *buf, size_t len, byte flags) {
+    int errcode;
+    const mp_stream_p_t *stream_out_p = ((mp_obj_base_t *)stream_out)->type->protocol;
+    mp_uint_t out_sz = stream_out_p->write(stream_out, buf, len, &errcode);
+    if (out_sz == MP_STREAM_ERROR) {
+        return mp_const_false;
+    }
+    return mp_const_none;
+}
+
+mp_uint_t mp_stream_read(mp_obj_t stream_in, void *buf, size_t len) {
+    int errcode;
+    const mp_stream_p_t *stream_in_p = ((mp_obj_base_t *)stream_in)->type->protocol;
+    mp_uint_t out_sz = stream_in_p->read(stream_in, buf, len, &errcode);
+    if (out_sz == MP_STREAM_ERROR) {
+        return -1;
+    }
+    return out_sz;
+}
+
+mp_uint_t mp_stream_poll(mp_obj_t stream_in, uintptr_t poll_flags) {
+    int errcode = 0;
+    mp_uint_t ret = 0;
+    const mp_stream_p_t *stream_in_p = ((mp_obj_base_t *)stream_in)->type->protocol;
+    ret = stream_in_p->ioctl(stream_in, MP_STREAM_POLL, poll_flags, &errcode);
+    return ret;
+}
+
+
+STATIC mp_obj_t rfcore_transparent(mp_obj_t stream_in, mp_obj_t stream_out, mp_obj_t callback) {
+    // Make sure we have suitable stream objects.
+    mp_get_stream_raise(stream_in, MP_STREAM_OP_READ | MP_STREAM_OP_IOCTL);
+    mp_get_stream_raise(stream_out, MP_STREAM_OP_WRITE);
+
+    mp_obj_t stm_module = mp_import_name(MP_QSTR_stm, mp_const_none, MP_OBJ_NEW_SMALL_INT(0));
+    mp_obj_t rfcore_ble_hci_obj = mp_import_from(stm_module, MP_QSTR_rfcore_ble_hci);
+
+    mp_obj_t time_module = mp_import_name(MP_QSTR_time, mp_const_none, MP_OBJ_NEW_SMALL_INT(0));
+    mp_obj_t sleep_ms_obj = mp_import_from(time_module, MP_QSTR_sleep_ms);
+
+    uint8_t buf[1024];
+    size_t rx = 0;
+    size_t len = 0;
+    int state = 0;
+    int cmd_type = 0;
+
+    while (true) {
+        // Sleep briefly to allow micropython background processing.
+        mp_call_function_n_kw(sleep_ms_obj, 1, 0, (mp_obj_t[]) {MP_OBJ_NEW_SMALL_INT(1)});
```

Comment:
I feel like this function should just do one iteration and then return to Python.  Then it could, for example, be used with asyncio.

But that's a pretty big change, so I don't expect it to be made.  Could be done in the future if anyone wanted to do it.